monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,222 @@
+
+BINARY_OS_MAPPING_KEY = "binaries"
+
+logger = logging.getLogger(__name__)
+
+
+class MongoAgentPluginRepository(IAgentPluginRepository):
+    def __init__(self, mongo_client: MongoClient) -> None:
+        self._agent_plugins_collection = mongo_client.monkey_island.agent_plugins
+        self._agent_plugins_binaries_collections = self._get_binary_collections(mongo_client)
+
+    def _get_binary_collections(
+        self, mongo_client: MongoClient
+    ) -> Dict[OperatingSystem, gridfs.GridFS]:
+        agent_plugins_binaries_collections: Dict[OperatingSystem, gridfs.GridFS] = {}
+
+        for os in OperatingSystem:
+            agent_plugins_binaries_collections[os] = gridfs.GridFS(
+                mongo_client.monkey_island, self._get_binary_collection_name(os)
+            )
+
+        return agent_plugins_binaries_collections
+
+    def _get_binary_collection_name(self, operating_system: OperatingSystem) -> str:
+        return f"agent_plugins_binaries_{operating_system.value}"
+
+    def get_plugin(
+        self, host_operating_system: OperatingSystem, plugin_type: AgentPluginType, name: PluginName
+    ) -> AgentPlugin:
+        try:
+            plugin_dict = self._get_agent_plugin(plugin_type, name)
+            os_binaries = plugin_dict[BINARY_OS_MAPPING_KEY]
+            gridfs_file_id = os_binaries[host_operating_system.value]
+            plugin_source_bytes = (
+                self._agent_plugins_binaries_collections[host_operating_system]
+                .get(gridfs_file_id)
+                .read()
+            )
+
+            plugin_dict.pop(BINARY_OS_MAPPING_KEY)
+            plugin_dict["source_archive"] = plugin_source_bytes
+            return AgentPlugin(**plugin_dict)
+        except UnknownRecordError:
+            raise
+        except Exception as err:
+            raise RetrievalError(
+                f"Error retrieving the agent plugin {name} of type {plugin_type} for operating "
+                f"system {host_operating_system}: {err}"
+            )
+
+    def _get_agent_plugin(
+        self, plugin_type: AgentPluginType, plugin_name: PluginName
+    ) -> Dict[str, Any]:
+        plugin_dict = self._agent_plugins_collection.find_one(
+            {
+                "plugin_manifest.name": plugin_name,
+                "plugin_manifest.plugin_type": plugin_type.value,
+            },
+            {MONGO_OBJECT_ID_KEY: False},
+        )
+
+        if plugin_dict is None:
+            raise UnknownRecordError(
+                f"Error retrieving the agent plugin {plugin_name} of type {plugin_type}"
+            )
+
+        return plugin_dict
+
+    def get_all_plugin_configuration_schemas(
+        self,
+    ) -> Dict[AgentPluginType, Dict[PluginName, Dict[str, Any]]]:
+        configuration_schema_dicts = self._agent_plugins_collection.find(
+            {}, {"plugin_manifest.plugin_type": 1, "plugin_manifest.name": 1, "config_schema": 1}
+        )
```

Comment:
If this call raises an exception (connection issue or any other), it needs to be wrapped in a RetrievalError.

Reply:
I'm not a fan of this approach. We can wrap every mongo call, but if there's no connection then it won't help